### DISCUSSION TITLE
Adding Ollama integration

### DISCUSSION THREAD

**Discussion by perfectra1n** [soulsands/trilium-chat#20] [General]
Hey all,

I added a "form" of Ollama integration that works via just a simple request that the plugin then displays. It doesn't do anything fancier, and is mainly just an (ab)use of the existing code to make it think it's talking to ChatGPT using OpenAI's API as the existing codebase is very tightly coupled to it. It would be great to see a new "Ollama" engine within the codebase, or to allow streaming!

If someone wants to integrate a new Ollama engine or something of the sort, a PR would certainly be accepted!

**Comment by j13055** [comment]
Not sure if this is the right spot for this question, but i have a problem with my setup and getting responses from the API-Request, when accessing Trilium from local network.
My Setup:
Created a docker compose file for creating Trilium (port 3030) and Ollama (port 11434) together in a docker container. All of this works fine.
The plugin itselfs works fine too and i get responses when using trilium on the machine where docker is running.
![image](https://github.com/user-attachments/assets/8f97e417-51d4-49b5-8149-374bde6bc4c3)


My Problem is that when i access my trilium instance from another pc in local network via the ip and correct port, i can use trilium completly fine, but getting following error when trying to chat with ollama in trilium.
![image](https://github.com/user-attachments/assets/2c270361-716c-4938-bd59-306271dd175b)
 
With these error codes:
```
>Access to fetch at 'http://-IP-OF-DOCKER-MACHINE-:11434/v1/chat/completions' from origin 'http://-IP-OF-DOCKER-MACHINE-:3030' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
VM30:5 
        
>POST http://-IP-OF-DOCKER-MACHINE-:11434/v1/chat/completions net::ERR_FAILED

>VM30:5 
TypeError: Failed to fetch
    at c.requestCompletionBatch (eval at <anonymous> (bundle.js:21:20), <anonymous>:5:31364)
    at c.requestChatBatch (eval at <anonymous> (bundle.js:21:20), <anonymous>:5:31256)
    at c.sendRequest (eval at <anonymous> (bundle.js:21:20), <anonymous>:5:29503)
    at c.requestCompletion (eval at <anonymous> (bundle.js:21:20), <anonymous>:5:29250)
    at E.t (eval at <anonymous> (bundle.js:21:20), <anonymous>:5:39634)
```
Where do i need to set things up correctly to allow it in local network?

**Comment by perfectra1n** [comment]
Hey there @j13055, the most important part of that error is the following:

> been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

This is common in browsers and Electron apps (like Trilium). [Here](https://github.com/ollama/ollama/issues/669#issuecomment-1783903239) is the ollama/ollama issue that outlines the fix, which is to modify the `OLLAMA_ORIGINS` environment variable on your ollama container. You can set it to `*` if you would like to allow all traffic to your `ollama`.

**Reply by perfectra1n** [reply]
I've also gone ahead and updated the [README.md](https://github.com/soulsands/trilium-chat/commit/d897a36e1f3c6927fdc950da774d208debb0e480) with these suggestions for future users. Thanks for pointing that out!

**Reply by j13055** [reply]
Thanks. I'll try that out :+1:

**Reply by j13055** [reply]
It worked. Thank you.

**Reply by perfectra1n** [reply]
Anytime!
